File: backend/api/custom_synonyms.py
"""
Custom organization synonyms management endpoints.

This module provides endpoints for managing organization-specific custom skill synonyms,
including CRUD operations for creating, reading, updating, and deleting custom synonym
mappings with context information.
"""
import logging
from typing import List, Optional

from fastapi import APIRouter, HTTPException, Query, status
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field
from sqlalchemy.exc import SQLAlchemyError

from models.custom_synonyms import CustomSynonym

logger = logging.getLogger(__name__)

router = APIRouter()


class CustomSynonymEntry(BaseModel):
    """Individual custom synonym entry definition."""

    canonical_skill: str = Field(..., description="Canonical name of the skill")
    custom_synonyms: List[str] = Field(..., description="Organization-specific synonyms for this skill")
    context: Optional[str] = Field(None, description="Context category (e.g., web_framework, language, database)")
    is_active: bool = Field(True, description="Whether this synonym mapping is currently active")


class CustomSynonymCreate(BaseModel):
    """Request model for creating custom synonyms."""

    organization_id: str = Field(..., description="Organization identifier")
    created_by: Optional[str] = Field(None, description="User ID who is creating these synonyms")
    synonyms: List[CustomSynonymEntry] = Field(..., description="List of custom synonym entries to create")


class CustomSynonymUpdate(BaseModel):
    """Request model for updating a custom synonym."""

    canonical_skill: Optional[str] = Field(None, description="Canonical name of the skill")
    custom_synonyms: Optional[List[str]] = Field(None, description="Organization-specific synonyms")
    context: Optional[str] = Field(None, description="Context category")
    is_active: Optional[bool] = Field(None, description="Whether this synonym mapping is active")


class CustomSynonymResponse(BaseModel):
    """Response model for a single custom synonym entry."""

    id: str = Field(..., description="Unique identifier for the synonym entry")
    organization_id: str = Field(..., description="Organization identifier")
    canonical_skill: str = Field(..., description="Canonical name of the skill")
    custom_synonyms: List[str] = Field(..., description="Organization-specific synonyms")
    context: Optional[str] = Field(None, description="Context category")
    created_by: Optional[str] = Field(None, description="User ID who created this entry")
    is_active: bool = Field(..., description="Whether this synonym mapping is active")
    created_at: str = Field(..., description="Creation timestamp")
    updated_at: str = Field(..., description="Last update timestamp")


class CustomSynonymListResponse(BaseModel):
    """Response model for listing custom synonyms."""

    organization_id: str = Field(..., description="Organization identifier")
    synonyms: List[CustomSynonymResponse] = Field(..., description="List of custom synonym entries")
    total_count: int = Field(..., description="Total number of entries")


@router.post(
    "/",
    response_model=CustomSynonymListResponse,
    status_code=status.HTTP_201_CREATED,
    tags=["Custom Synonyms"],
)
async def create_custom_synonyms(request: CustomSynonymCreate) -> JSONResponse:
    """
    Create custom synonym entries for an organization.

    This endpoint accepts a batch of custom synonym entries for a specific organization,
    validating the data and creating database records for each skill with its custom
    synonyms and context information.

    Args:
        request: Create request with organization_id and list of synonyms

    Returns:
        JSON response with created synonym entries

    Raises:
        HTTPException(422): If validation fails
        HTTPException(500): If database operation fails

    Examples:
        >>> import requests
        >>> data = {
        ...     "organization_id": "org123",
        ...     "created_by": "user456",
        ...     "synonyms": [
        ...         {
        ...             "canonical_skill": "React",
        ...             "context": "web_framework",
        ...             "custom_synonyms": ["ReactJS", "React.js", "React Framework"],
        ...             "is_active": True
        ...         }
        ...     ]
        ... }
        >>> response = requests.post("http://localhost:8000/api/custom-synonyms/", json=data)
        >>> response.json()
        {
            "organization_id": "org123",
            "synonyms": [...],
            "total_count": 1
        }
    """
    try:
        logger.info(f"Creating {len(request.synonyms)} custom synonyms for organization: {request.organization_id}")

        # Validate organization_id
        if not request.organization_id or len(request.organization_id.strip()) == 0:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail="Organization ID cannot be empty",
            )

        # Validate synonyms list
        if not request.synonyms or len(request.synonyms) == 0:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail="At least one synonym entry must be provided",
            )

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        created_synonyms = []
        for synonym in request.synonyms:
            # Placeholder synonym entry
            synonym_response = {
                "id": "placeholder-id",
                "organization_id": request.organization_id,
                "canonical_skill": synonym.canonical_skill,
                "custom_synonyms": synonym.custom_synonyms,
                "context": synonym.context,
                "created_by": request.created_by,
                "is_active": synonym.is_active,
                "created_at": "2024-01-25T00:00:00Z",
                "updated_at": "2024-01-25T00:00:00Z",
            }
            created_synonyms.append(synonym_response)

        response_data = {
            "organization_id": request.organization_id,
            "synonyms": created_synonyms,
            "total_count": len(created_synonyms),
        }

        logger.info(f"Created {len(created_synonyms)} custom synonyms for organization: {request.organization_id}")

        return JSONResponse(
            status_code=status.HTTP_201_CREATED,
            content=response_data,
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error creating custom synonyms: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to create custom synonyms: {str(e)}",
        ) from e


@router.get("/", tags=["Custom Synonyms"])
async def list_custom_synonyms(
    organization_id: Optional[str] = Query(None, description="Filter by organization ID"),
    canonical_skill: Optional[str] = Query(None, description="Filter by canonical skill name"),
    is_active: Optional[bool] = Query(None, description="Filter by active status"),
) -> JSONResponse:
    """
    List custom synonym entries with optional filters.

    Args:
        organization_id: Optional organization ID filter
        canonical_skill: Optional canonical skill filter
        is_active: Optional active status filter

    Returns:
        JSON response with list of custom synonym entries

    Raises:
        HTTPException(500): If database query fails

    Examples:
        >>> import requests
        >>> response = requests.get("http://localhost:8000/api/custom-synonyms/?organization_id=org123")
        >>> response.json()
    """
    try:
        logger.info(f"Listing custom synonyms with filters - organization_id: {organization_id}, canonical_skill: {canonical_skill}, is_active: {is_active}")

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        response_data = {
            "organization_id": organization_id or "all",
            "synonyms": [],
            "total_count": 0,
        }

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=response_data,
        )

    except Exception as e:
        logger.error(f"Error listing custom synonyms: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to list custom synonyms: {str(e)}",
        ) from e


@router.get("/{synonym_id}", tags=["Custom Synonyms"])
async def get_custom_synonym(synonym_id: str) -> JSONResponse:
    """
    Get a specific custom synonym entry by ID.

    Args:
        synonym_id: Unique identifier of the synonym entry

    Returns:
        JSON response with synonym entry details

    Raises:
        HTTPException(404): If synonym entry is not found
        HTTPException(500): If database query fails

    Examples:
        >>> import requests
        >>> response = requests.get("http://localhost:8000/api/custom-synonyms/123e4567-e89b-12d3-a456-426614174000")
        >>> response.json()
    """
    try:
        logger.info(f"Getting custom synonym: {synonym_id}")

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "id": synonym_id,
                "organization_id": "org123",
                "canonical_skill": "React",
                "custom_synonyms": ["ReactJS", "React.js", "React Framework"],
                "context": "web_framework",
                "created_by": "user456",
                "is_active": True,
                "created_at": "2024-01-25T00:00:00Z",
                "updated_at": "2024-01-25T00:00:00Z",
            },
        )

    except Exception as e:
        logger.error(f"Error getting custom synonym: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get custom synonym: {str(e)}",
        ) from e


@router.put("/{synonym_id}", tags=["Custom Synonyms"])
async def update_custom_synonym(
    synonym_id: str, request: CustomSynonymUpdate
) -> JSONResponse:
    """
    Update a custom synonym entry.

    Args:
        synonym_id: Unique identifier of the synonym entry
        request: Update request with fields to modify

    Returns:
        JSON response with updated synonym entry

    Raises:
        HTTPException(404): If synonym entry is not found
        HTTPException(422): If validation fails
        HTTPException(500): If database operation fails

    Examples:
        >>> import requests
        >>> data = {"custom_synonyms": ["React", "ReactJS", "React.js", "React Framework"]}
        >>> response = requests.put(
        ...     "http://localhost:8000/api/custom-synonyms/123",
        ...     json=data
        ... )
        >>> response.json()
    """
    try:
        logger.info(f"Updating custom synonym: {synonym_id}")

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "id": synonym_id,
                "organization_id": "org123",
                "canonical_skill": request.canonical_skill or "React",
                "custom_synonyms": request.custom_synonyms or ["React", "ReactJS"],
                "context": request.context,
                "created_by": "user456",
                "is_active": request.is_active if request.is_active is not None else True,
                "created_at": "2024-01-25T00:00:00Z",
                "updated_at": "2024-01-25T00:00:00Z",
            },
        )

    except Exception as e:
        logger.error(f"Error updating custom synonym: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to update custom synonym: {str(e)}",
        ) from e


@router.delete("/{synonym_id}", tags=["Custom Synonyms"])
async def delete_custom_synonym(synonym_id: str) -> JSONResponse:
    """
    Delete a custom synonym entry.

    Args:
        synonym_id: Unique identifier of the synonym entry

    Returns:
        JSON response confirming deletion

    Raises:
        HTTPException(404): If synonym entry is not found
        HTTPException(500): If database operation fails

    Examples:
        >>> import requests
        >>> response = requests.delete("http://localhost:8000/api/custom-synonyms/123")
        >>> response.json()
        {"message": "Custom synonym deleted successfully"}
    """
    try:
        logger.info(f"Deleting custom synonym: {synonym_id}")

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={"message": f"Custom synonym {synonym_id} deleted successfully"},
        )

    except Exception as e:
        logger.error(f"Error deleting custom synonym: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to delete custom synonym: {str(e)}",
        ) from e


@router.delete("/organization/{organization_id}", tags=["Custom Synonyms"])
async def delete_custom_synonyms_by_organization(organization_id: str) -> JSONResponse:
    """
    Delete all custom synonym entries for a specific organization.

    Args:
        organization_id: Organization identifier to delete synonyms for

    Returns:
        JSON response confirming deletion with count

    Raises:
        HTTPException(500): If database operation fails

    Examples:
        >>> import requests
        >>> response = requests.delete("http://localhost:8000/api/custom-synonyms/organization/org123")
        >>> response.json()
        {"message": "Deleted 5 custom synonyms for organization: org123"}
    """
    try:
        logger.info(f"Deleting all custom synonyms for organization: {organization_id}")

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={"message": f"Deleted custom synonyms for organization: {organization_id}", "deleted_count": 0},
        )

    except Exception as e:
        logger.error(f"Error deleting custom synonyms for organization {organization_id}: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to delete custom synonyms: {str(e)}",
        ) from e

File: frontend/src/api/customSynonyms.ts
/**
 * Custom Synonyms API Client
 *
 * This module provides a client for managing custom synonym CRUD operations.
 * Supports creating, listing, retrieving, updating, and deleting custom synonyms
 * for organizations.
 *
 * @example
 * ```ts
 * import { customSynonymsClient } from '@/api/customSynonyms';
 *
 * // Create new custom synonyms for an organization
 * const created = await customSynonymsClient.createCustomSynonyms({
 *   organization_id: 'org123',
 *   created_by: 'user456',
 *   synonyms: [
 *     {
 *       canonical_skill: 'React',
 *       custom_synonyms: ['ReactJS', 'React.js', 'React Framework'],
 *       context: 'web_framework',
 *       is_active: true,
 *     },
 *   ],
 * });
 *
 * // List all custom synonyms for an organization
 * const list = await customSynonymsClient.listCustomSynonyms('org123');
 *
 * // Get a specific custom synonym entry
 * const synonym = await customSynonymsClient.getCustomSynonym('synonym-123');
 *
 * // Update a custom synonym
 * const updated = await customSynonymsClient.updateCustomSynonym('synonym-123', {
 *   canonical_skill: 'React',
 *   custom_synonyms: ['ReactJS', 'React.js'],
 *   is_active: true,
 * });
 *
 * // Delete a custom synonym
 * await customSynonymsClient.deleteCustomSynonym('synonym-123');
 *
 * // Delete all custom synonyms for an organization
 * await customSynonymsClient.deleteCustomSynonymsByOrganization('org123');
 * ```
 */

import axios, { AxiosInstance, AxiosError, AxiosResponse } from 'axios';
import { config } from '@/config';
import type {
  CustomSynonymCreate,
  CustomSynonymUpdate,
  CustomSynonymResponse,
  CustomSynonymListResponse,
  ApiError,
} from '@/types/api';

/**
 * Default API configuration for custom synonyms client
 */
const DEFAULT_CONFIG = {
  baseURL: config.api.url,
  timeout: 10000, // 10 seconds
  headers: {
    'Content-Type': 'application/json',
  },
};

/**
 * List options for custom synonyms
 */
export interface ListCustomSynonymsOptions {
  organization_id?: string;
  canonical_skill?: string;
  is_active?: boolean;
}

/**
 * Custom Synonyms API Client class
 *
 * Provides methods for managing custom synonyms with proper error handling
 * and type safety.
 */
export class CustomSynonymsClient {
  private client: AxiosInstance;

  /**
   * Create a new Custom Synonyms client instance
   *
   * @param configOverride - Optional configuration overrides
   */
  constructor(configOverride: Partial<typeof DEFAULT_CONFIG> = {}) {
    const finalConfig = { ...DEFAULT_CONFIG, ...configOverride };

    this.client = axios.create(finalConfig);

    // Response interceptor for error handling
    this.client.interceptors.response.use(
      (response) => response,
      (error) => Promise.reject(this.transformError(error))
    );
  }

  /**
   * Transform Axios error to standardized API error
   *
   * @param error - Axios error
   * @returns Transformed API error
   */
  private transformError(error: unknown): ApiError {
    const axiosError = error as AxiosError<{ detail?: string }>;

    // Network error (no response)
    if (!axiosError.response) {
      if (axiosError.code === 'ECONNABORTED') {
        return {
          detail: 'Request timeout. Please check your connection and try again.',
          status: 408,
        };
      }
      return {
        detail: 'Network error. Please check your connection and try again.',
        status: 0,
      };
    }

    // Server returned error response
    const status = axiosError.response.status;
    const data = axiosError.response.data;

    // Use server's error message if available
    if (data?.detail) {
      return { detail: data.detail, status };
    }

    // Default error messages by status code
    const defaultMessages: Record<number, string> = {
      400: 'Invalid request. Please check your input.',
      401: 'Unauthorized. Please log in.',
      403: 'Forbidden. You do not have permission.',
      404: 'Resource not found.',
      422: 'Validation error. Please check your input.',
      429: 'Too many requests. Please try again later.',
      500: 'Server error. Please try again later.',
      502: 'Bad gateway. Please try again later.',
      503: 'Service unavailable. Please try again later.',
    };

    return {
      detail: data?.detail || defaultMessages[status] || 'An unexpected error occurred.',
      status,
    };
  }

  /**
   * Create custom synonym entries for an organization
   *
   * @param data - Custom synonym creation request
   * @returns Created synonym entries
   * @throws ApiError if creation fails
   *
   * @example
   * ```ts
   * const result = await customSynonymsClient.createCustomSynonyms({
   *   organization_id: 'org123',
   *   created_by: 'user456',
   *   synonyms: [
   *     {
   *       canonical_skill: 'React',
   *       custom_synonyms: ['ReactJS', 'React.js', 'React Framework'],
   *       context: 'web_framework',
   *       is_active: true,
   *     },
   *   ],
   * });
   * ```
   */
  async createCustomSynonyms(
    data: CustomSynonymCreate
  ): Promise<CustomSynonymListResponse> {
    try {
      const response: AxiosResponse<CustomSynonymListResponse> =
        await this.client.post('/api/custom-synonyms/', data);
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * List custom synonyms with optional filters
   *
   * @param options - List options including organization_id, canonical_skill, and is_active filters
   * @returns List of custom synonym entries
   * @throws ApiError if listing fails
   *
   * @example
   * ```ts
   * // Get all custom synonyms for an organization
   * const list = await customSynonymsClient.listCustomSynonyms({
   *   organization_id: 'org123',
   * });
   *
   * // Filter by canonical skill and active status
   * const filtered = await customSynonymsClient.listCustomSynonyms({
   *   organization_id: 'org123',
   *   canonical_skill: 'React',
   *   is_active: true,
   * });
   * ```
   */
  async listCustomSynonyms(
    options: ListCustomSynonymsOptions = {}
  ): Promise<CustomSynonymListResponse[]> {
    try {
      const { organization_id, canonical_skill, is_active } = options;

      const params: Record<string, string | boolean> = {};

      if (organization_id) {
        params.organization_id = organization_id;
      }
      if (canonical_skill) {
        params.canonical_skill = canonical_skill;
      }
      if (is_active !== undefined) {
        params.is_active = is_active;
      }

      const response: AxiosResponse<CustomSynonymListResponse[]> =
        await this.client.get('/api/custom-synonyms/', { params });
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Get a specific custom synonym entry by ID
   *
   * @param id - Synonym entry ID
   * @returns Custom synonym entry
   * @throws ApiError if not found
   *
   * @example
   * ```ts
   * const synonym = await customSynonymsClient.getCustomSynonym('synonym-123');
   * console.log(synonym.canonical_skill, synonym.custom_synonyms);
   * ```
   */
  async getCustomSynonym(id: string): Promise<CustomSynonymResponse> {
    try {
      const response: AxiosResponse<CustomSynonymResponse> =
        await this.client.get(`/api/custom-synonyms/${id}`);
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Update a custom synonym entry
   *
   * @param id - Synonym entry ID
   * @param data - Update request
   * @returns Updated synonym entry
   * @throws ApiError if update fails
   *
   * @example
   * ```ts
   * const updated = await customSynonymsClient.updateCustomSynonym(
   *   'synonym-123',
   *   {
   *     canonical_skill: 'React',
   *     custom_synonyms: ['ReactJS', 'React.js'],
   *     is_active: true,
   *   }
   * );
   * ```
   */
  async updateCustomSynonym(
    id: string,
    data: CustomSynonymUpdate
  ): Promise<CustomSynonymResponse> {
    try {
      const response: AxiosResponse<CustomSynonymResponse> =
        await this.client.put(`/api/custom-synonyms/${id}`, data);
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Delete a specific custom synonym entry
   *
   * @param id - Synonym entry ID
   * @throws ApiError if deletion fails
   *
   * @example
   * ```ts
   * await customSynonymsClient.deleteCustomSynonym('synonym-123');
   * ```
   */
  async deleteCustomSynonym(id: string): Promise<void> {
    try {
      await this.client.delete(`/api/custom-synonyms/${id}`);
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Delete all custom synonyms for an organization
   *
   * @param organizationId - Organization ID
   * @throws ApiError if deletion fails
   *
   * @example
   * ```ts
   * await customSynonymsClient.deleteCustomSynonymsByOrganization('org123');
   * ```
   */
  async deleteCustomSynonymsByOrganization(organizationId: string): Promise<void> {
    try {
      await this.client.delete(`/api/custom-synonyms/organization/${organizationId}`);
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Get the underlying Axios instance
   *
   * This is useful for making custom requests not covered by the convenience methods.
   *
   * @returns Axios instance
   */
  getAxiosInstance(): AxiosInstance {
    return this.client;
  }
}

/**
 * Default custom synonyms client instance
 *
 * Use this singleton instance for all custom synonym operations.
 */
export const customSynonymsClient = new CustomSynonymsClient();

/**
 * Export custom synonyms client class for custom instances
 */
export default CustomSynonymsClient;
